File: SWEA/D4/1868. 파핑파핑 지뢰찾기/파핑파핑 지뢰찾기.py
from collections import deque

def countBomb(x, y):
    cnt = 0

    for i in range(8):
        nx = x + dx[i]
        ny = y + dy[i]

        if nx < 0 or nx >= n or ny < 0 or ny >= n:
            continue

        if matrix[nx][ny] == '*':
            cnt += 1

    matrix[x][y] = cnt


def punkCnt(x, y):

    global cnt
    q = deque()
    q.append([x, y])
    matrix[x][y] = '*'

    while q:
        a, b = q.popleft()

        for i in range(8):
            nx = a + dx[i]
            ny = b + dy[i]

            if nx < 0 or nx >= n or ny < 0 or ny >= n:
                continue

            if matrix[nx][ny] == 0:
                matrix[nx][ny] = '*'
                q.append([nx, ny])

            elif matrix[nx][ny] != '*':
                matrix[nx][ny] = '*'
                cnt += 1


T = int(input())
dx, dy = [-1, 1, 0, 0, 1, 1, -1, -1], [0, 0, -1, 1, 1, -1, 1, -1]

for t in range(1, T + 1):

    n = int(input())
    matrix = [list(input()) for _ in range(n)]

    for i in range(n):
        for j in range(n):
            if matrix[i][j] == '.':
                countBomb(i, j)

    # for i in range(n):
    #     print(*matrix[i], end=' ')
    #     print()

    cnt = 0
    tmp = 0
    for i in range(n):
        for j in range(n):
            if matrix[i][j] == 0:
                punkCnt(i, j)
                tmp += 1

    cnt = 0
    for i in range(n):
        for j in range(n):
            if matrix[i][j] != '*':
                cnt += 1


    print(f"#{t} {tmp + cnt}")

File: SWEA/D4/1868. 파핑파핑 지뢰찾기/파핑파핑 지뢰찾기.java
import java.util.*;

class Solution {
    static int n;
    static char[][] matrix;
    static int[] dx = {-1, 1, 0, 0, 1, 1, -1, -1};
    static int[] dy = {0, 0, -1, 1, 1, -1, 1, -1};
    static int cnt;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();

        for (int t = 1; t <= T; t++) {
            n = sc.nextInt();
            matrix = new char[n][n];
            for (int i = 0; i < n; i++) {
                String line = sc.next();
                for (int j = 0; j < n; j++) {
                    matrix[i][j] = line.charAt(j);
                }
            }

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (matrix[i][j] == '.') {
                        countBomb(i, j);
                    }
                }
            }

            int tmp = 0;
            cnt = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (matrix[i][j] == '0') {
                        punkCnt(i, j);
                        tmp++;
                    }
                }
            }

            cnt = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (matrix[i][j] != '*') {
                        cnt++;
                    }
                }
            }

            System.out.printf("#%d %d\n", t, tmp + cnt);
        }
    }

    static void countBomb(int x, int y) {
        int bombCount = 0;
        for (int i = 0; i < 8; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx >= 0 && nx < n && ny >= 0 && ny < n && matrix[nx][ny] == '*') {
                bombCount++;
            }
        }
        matrix[x][y] = (char) ('0' + bombCount);
    }

    static void punkCnt(int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});
        matrix[x][y] = '*';

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int a = current[0];
            int b = current[1];

            for (int i = 0; i < 8; i++) {
                int nx = a + dx[i];
                int ny = b + dy[i];

                if (nx >= 0 && nx < n && ny >= 0 && ny < n) {
                    if (matrix[nx][ny] == '0') {
                        matrix[nx][ny] = '*';
                        q.add(new int[]{nx, ny});
                    } else if (matrix[nx][ny] != '*') {
                        matrix[nx][ny] = '*';
                        cnt++;
                    }
                }
            }
        }
    }
}